Middleware Fails when called from gateway

With this config:

```
<?xml version="1.0" encoding="UTF-8"?>
<middleware xmlns="urn:katana:middleware" name="api" version="1.0.0" protocol="urn:katana:protocol:http" request="true" response="true">
    <info title="Auth" description="Response middleware for API requests and responses" terms="https://tech.example.com/docs/middleware-api">
        <contact name="API" url="https://tech.example.com/docs/middleware-apii" email="api@api.com"/>
        <license name="License" url="https://tech.example.com/docs/middleware-api"/>
    </info>
    <external-docs description="API middleware docs" url="https://tech.example.com/docs/middleware-api"/>
    <engine runner="urn:katana:runner:node" path="../runner.js"/>
    <http-response code="401" description="The access credentials were invalid" mime="application/problem+json">
        <http-example>
            <![CDATA[
            {
              "type": "https://tech.example.com/api/errors#401-Unauthorized",
              "detail": "The access credentials used on this request are not valid",
              "status": 401,
              "title": "Unauthorized"
            }
            ]]>
        </http-example>
    </http-response>
</middleware>
```

And this code:
```
const Middleware = require('katana.sdk').Middleware;
const middleware = new Middleware();

middleware.request((request) => {
    return request;
});

middleware.response((response) => {
    return response;
});

middleware.run();
```

The gateway calls return this error on any type of request:
```
2017-06-09T07:58:40.104Z [ERROR] [SDK] Error running callback: response. Cannot read property 'v' of undefined TypeError: Cannot read property 'v' of undefined
at Mapper.getHttpRequest (/home/vagrant/dev/katana/middleware-api/node_modules/katana.sdk/sdk/mapper.js:101:20)
at Middleware._getResponse (/home/vagrant/dev/katana/middleware-api/node_modules/katana.sdk/sdk/component.js:234:20)
at Middleware._processCommand (/home/vagrant/dev/katana/middleware-api/node_modules/katana.sdk/sdk/middleware.js:30:33)
at Middleware._processMessage (/home/vagrant/dev/katana/middleware-api/node_modules/katana.sdk/sdk/component.js:142:10)
at emitThree (events.js:116:13)
at emit (events.js:194:7)
at Socket._flushRead (/home/vagrant/dev/katana/middleware-api/node_modules/zmq/lib/index.js:638:15)
at Socket._flushReads (/home/vagrant/dev/katana/middleware-api/node_modules/zmq/lib/index.js:676:23)
at _zmq.onReadReady (/home/vagrant/dev/katana/middleware-api/node_modules/zmq/lib/index.js:297:10)
```

If I run ```katana middleware request -c config/katana.xml -u http://test.com/1.0/auth/facebook``` it works as expected.

@jeronimoalbi @orestes any updates on this? This is causing important issues since the behaviour for middlewares is not stable.
@orestes I think this is SDK related, the following message is always shown when trying to retrieve the Service Schema from a middleware:

```
2017-07-19T18:28:44.628Z [ERROR] [SDK] Error running callback: request. Cannot get service schema. No mapping provided Error: Cannot get service schema. No mapping provided
```

I've debugged this line https://github.com/kusanagi/katana-sdk-node/blob/master/sdk/api.js#L157 and the **servicesMapping** property is always empty. Even if the services are running and registered.

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "katana.sdk",
-  "version": "1.1.0",
+  "version": "1.1.1",
   "description": "KATANA SDK for Node.js v6",
   "main": "index.js",
   "scripts": {

--- a/package.json
+++ b/package.json
@@ -49,12 +49,12 @@
     "immutable": "^3.8.1",
     "lodash": "^4.16.2",
     "msgpack": "^1.0.2",
+    "semver": "^5.3.0",
     "sync-request": "^3.0.1",
     "yargs": "^6.3.0",
     "zeromq": "^4.3.0"
   },
   "devDependencies": {
-    "np": "^2.13.1",
     "chai": "3.5.0",
     "commitizen": "^2.9.6",
     "coveralls": "2.12.0",
@@ -65,6 +65,7 @@
     "jsdoc": "^3.4.0",
     "mocha": "^3.1.2",
     "mocha-lcov-reporter": "1.3.0",
+    "np": "^2.13.1",
     "npm-run-all": "4.0.2",
     "rimraf": "^2.6.1",
     "sinon": "1.17.7",

--- a/sdk/action.js
+++ b/sdk/action.js
@@ -716,7 +716,7 @@ for action: "${this._actionName}"`
       (name) => this._params[name] instanceof Param ? this._params[name] : new Param(
         name,
         this._params[name][m.value],
-        this._params[m.type],
+        this._params[name][m.type],
         this.hasParam(name)
       )
     );

--- a/sdk/action.js
+++ b/sdk/action.js
@@ -252,32 +252,6 @@ for action: "${this._actionName}"`
   }
 
   /**
-   * @return {boolean}
-   */
-  hasReturn() {
-
-    try {
-      return this
-        .getServiceSchema(this._name, this._version)
-        .getActionSchema(this._actionName)
-        .hasReturn();
-    } catch (e) {
-      // If we don't have a schema, return a sensible default
-      return false;
-    }
-  }
-
-  /**
-   *
-   * @return {*}
-   */
-  getReturn() {
-    if (this._return) {
-      return this._return;
-    }
-  }
-
-  /**
    * @return {array|number|boolean|string}
    */
   getReturnType() {

--- a/sdk/api.js
+++ b/sdk/api.js
@@ -1,4 +1,5 @@
 'use strict';
+const semver = require('semver');
 
 /**
  * Base Api class
@@ -159,8 +160,11 @@ class Api {
     }
 
     const service = this._component.servicesMapping.find((definition) =>
-      name === definition.schema.getName() && version === definition.schema.getVersion()
+      name === definition.schema.getName()
+      &&
+      semver.satisfies(definition.schema.getVersion(), version)
     );
+    console.log(service);
 
     if (service) {
       return service.schema;

--- a/sdk/api.js
+++ b/sdk/api.js
@@ -164,7 +164,6 @@ class Api {
       &&
       semver.satisfies(definition.schema.getVersion(), version)
     );
-    console.log(service);
 
     if (service) {
       return service.schema;

--- a/sdk/api.test.js
+++ b/sdk/api.test.js
@@ -231,7 +231,7 @@ describe('Api', () => {
 
     const mockMapping = {
       'users': {
-        '1.0': {
+        '1.0.0': {
           'ac': {
             'list': {
               'D': false,
@@ -289,7 +289,7 @@ describe('Api', () => {
       const api = new Api(component);
 
       component._saveNewMapping(mockMapping);
-      expect(api.getServiceSchema('users', '1.0')).to.be.an.instanceOf(ServiceSchema);
+      expect(api.getServiceSchema('users', '1.0.*')).to.be.an.instanceOf(ServiceSchema);
     });
 
     it('returns an instance of ServiceSchema', () => {
@@ -297,8 +297,8 @@ describe('Api', () => {
       const api = new Api(component);
 
       component._saveNewMapping(mockMapping);
-      expect(api.getServiceSchema('users', '1.0').getName()).to.equal('users');
-      expect(api.getServiceSchema('users', '1.0').getVersion()).to.equal('1.0');
+      expect(api.getServiceSchema('users', '1.0.*').getName()).to.equal('users');
+      expect(api.getServiceSchema('users', '1.0.*').getVersion()).to.equal('1.0.0');
     });
 
   });

--- a/sdk/component.js
+++ b/sdk/component.js
@@ -252,7 +252,8 @@ class Component {
       this._mapper.getHttpResponse(payload[m.command][m.arguments][m.response]),
       this._mapper.getTransport(payload[m.command][m.arguments][m.transport]),
       payload[m.meta][m.protocol],
-      payload[m.meta][m.gateway]
+      payload[m.meta][m.gateway],
+      payload[m.command][m.arguments][m.return_value]
     );
   }
 

--- a/sdk/component.js
+++ b/sdk/component.js
@@ -248,9 +248,9 @@ class Component {
       this.cli.frameworkVersion,
       this.cli.variables,
       this.cli.debug,
-      this._mapper.getHttpRequest(payload[m.response]),
-      this._mapper.getHttpResponse(payload[m.response]),
-      this._mapper.getTransport(payload[m.transport]),
+      this._mapper.getHttpRequest(payload[m.command][m.arguments][m.request]),
+      this._mapper.getHttpResponse(payload[m.command][m.arguments][m.response]),
+      this._mapper.getTransport(payload[m.command][m.arguments][m.transport]),
       payload[m.meta][m.protocol],
       payload[m.meta][m.gateway]
     );

--- a/sdk/http-request.js
+++ b/sdk/http-request.js
@@ -192,11 +192,11 @@ class HttpRequest {
       throw new TypeError('The header `name` must be a string');
     }
 
-    return this[_data].getIn(['headers', name]) || defaultValue;
+    return this[_data].getIn(['headers', name], defaultValue);
   }
 
   getHeaders() {
-    return this[_data].get('headers').toJS();
+    return this[_data].has('headers') ? this[_data].get('headers').toJS() : {};
   }
 
   hasBody() {

--- a/sdk/http-request.js
+++ b/sdk/http-request.js
@@ -192,7 +192,10 @@ class HttpRequest {
       throw new TypeError('The header `name` must be a string');
     }
 
-    return this[_data].getIn(['headers', name], defaultValue);
+    return Immutable.List.isList(this[_data].getIn(['headers', name], defaultValue)) ?
+      this[_data].getIn(['headers', name], defaultValue).get(0)
+      :
+      this[_data].getIn(['headers', name], defaultValue);
   }
 
   getHeaders() {

--- a/sdk/http-request.test.js
+++ b/sdk/http-request.test.js
@@ -284,6 +284,15 @@ describe('HttpRequest', () => {
       assert.equal(httpRequest.getHeader('Content-Type'), 'test');
     });
 
+    it('should return the default value when header is not present', () => {
+      const headers      = {};
+      const _mockRequest = _.merge({headers}, mockRequest);
+      const httpRequest  = new HttpRequest(_mockRequest);
+      const defaultValue = 'text/plain';
+
+      assert.equal(httpRequest.getHeader('Content-Type', defaultValue), defaultValue);
+    });
+
     it('should throw an error if no header `name` specified', () => {
       const httpRequest = new HttpRequest(mockRequest);
       assert.throws(httpRequest.getHeader, /Specify a header `name`/);
@@ -296,6 +305,11 @@ describe('HttpRequest', () => {
       const _mockRequest = _.merge({headers}, mockRequest);
       const httpRequest = new HttpRequest(_mockRequest);
       assert.deepEqual(httpRequest.getHeaders(), headers);
+    });
+
+    it('should return an empty set of headers when none are set', () => {
+      const httpRequest = new HttpRequest(mockRequest);
+      assert.deepEqual(httpRequest.getHeaders(), {});
     });
   });
 

--- a/sdk/http-response.js
+++ b/sdk/http-response.js
@@ -10,7 +10,10 @@ const _statusText = Symbol('statusText');
 class HttpResponse {
   constructor(data) {
     this[_data] = Immutable.fromJS(data);
-    [this[_statusCode], this[_statusText]] = this[_data].get('status').split(' ');
+    const parts = this[_data].get('status').split(' ');
+
+    this[_statusCode] = parseInt(parts[0], 10);
+    this[_statusText] = parts.slice(1).join(' ');
   }
 
   isProtocolVersion(version) {

--- a/sdk/http-response.js
+++ b/sdk/http-response.js
@@ -95,7 +95,10 @@ class HttpResponse {
       throw new TypeError('The header `name` must be a string');
     }
 
-    return this[_data].getIn(['headers', name], defaultValue);
+    return Immutable.List.isList(this[_data].getIn(['headers', name], defaultValue)) ?
+      this[_data].getIn(['headers', name], defaultValue).get(0)
+      :
+      this[_data].getIn(['headers', name], defaultValue);
   }
 
   getHeaders() {

--- a/sdk/http-response.js
+++ b/sdk/http-response.js
@@ -88,14 +88,14 @@ class HttpResponse {
     return this[_data].hasIn(['headers', name]);
   }
 
-  getHeader(name) {
+  getHeader(name, defaultValue = '') {
     if (_.isNil(name)) {
       throw new Error('Specify a header `name`');
     } else if (!_.isString(name)) {
       throw new TypeError('The header `name` must be a string');
     }
 
-    return this[_data].getIn(['headers', name]) || '';
+    return this[_data].getIn(['headers', name], defaultValue);
   }
 
   getHeaders() {

--- a/sdk/http-response.test.js
+++ b/sdk/http-response.test.js
@@ -171,10 +171,14 @@ describe('HttpResponse', () => {
       assert.equal(httpResponse.getHeader('Content-Type'), 'test');
     });
 
-    it('should return an empty string if specified header does not exist', () => {
+
+    it('should return the default value if specified header does not exist', () => {
       const _mockResponse = Object.assign({headers: {}}, mockResponse);
-      const httpResponse = new HttpResponse(_mockResponse);
-      assert.equal(httpResponse.getHeader('Content-Type'), '');
+      const httpResponse  = new HttpResponse(_mockResponse);
+
+      const defaultValue = 'text/plain';
+
+      assert.equal(httpResponse.getHeader('Content-Type', defaultValue), defaultValue);
     });
 
     it('should throw an error if no header `name` specified', () => {
@@ -194,6 +198,11 @@ describe('HttpResponse', () => {
       const _mockResponse = Object.assign({headers}, mockResponse);
       const httpResponse = new HttpResponse(_mockResponse);
       assert.deepEqual(httpResponse.getHeaders(), headers);
+    });
+
+    it('should return an empty set of headers when none are set', () => {
+      const httpResponse = new HttpResponse(mockResponse);
+      assert.deepEqual(httpResponse.getHeaders(), {});
     });
   });
 

--- a/sdk/mapper.js
+++ b/sdk/mapper.js
@@ -140,7 +140,10 @@ class Mapper {
   }
 
   getHttpResponse(data) {
-    // const responseData = data[m.command][m.arguments][m.response];
+
+    if (!data) {
+      throw new Error('Cannot get HTTP response without data');
+    }
 
     const httpResponseData = {
       version: data[m.version],

--- a/sdk/mapper.test.js
+++ b/sdk/mapper.test.js
@@ -65,7 +65,9 @@ describe('mapper', () => {
       const mapper = new Mapper();
       const transport = mapper.getTransport(mockRequestTransport[m.command][m.arguments]);
       const httpResponse = new HttpResponse(mockResponse);
-      const resp = new Response(null, '', '', '', null, {}, false, null, httpResponse, transport);
+      const resp = new Response(
+        null, '', '', '', '', {}, false, null, httpResponse, transport, '', '', null
+      );
       const mock = sinon.mock(resp);
       mock.expects('hasReturn').once().returns(true);
       mock.expects('getReturnType').once().returns(null);

--- a/sdk/mappings.js
+++ b/sdk/mappings.js
@@ -83,6 +83,7 @@ const Mappings = {
   'required': 'r',
   'response': 'R',
   'result': 'r',
+  'return_value': 'rv',
   'response_return': 'rv',
   'return': 'R',
   'schema': 's',

--- a/sdk/middleware.test.js
+++ b/sdk/middleware.test.js
@@ -105,46 +105,63 @@ describe('Middleware', () => {
       });
 
       const response = {
-        'm': {
-          'v': '1.0.0',
-          'i': 'f1b27da9-240b-40e3-99dd-a567e4498ed7',
-          'd': '2016-04-12T02:49:05.761',
-          't': 2,
-          'p': 'http',
-          'g': ['12.34.56.78:1234', 'http://127.0.0.1:80']
-        },
-        'r': {
-          'v': '1.1',
-          'm': 'POST',
-          'u': 'http://example.com/v1.0.0/users',
-          'q': {
-            'name': ['James'],/**/
-            'age': ['32']
-          },
-          'h': {
-            'Accept': 'application/json'
-          },
-          'b': ''
-        },
-        'R': {
-          'v': '1.1',
-          's': '200 OK',
-          'h': {
-            'Content-Type': 'application/json'
-          },
-          'b': '{"result":"OK","message":"Created new user"}'
-        },
-        'T': {
-          'm': {
-            'v': '1.0.0',
-            'i': 'f1b27da9-240b-40e3-99dd-a567e4498ed7',
-            'd': '2016-04-12T02:49:05.761',
-            'g': ['12.34.56.78:1234', 'http://127.0.0.1:80'],
-            'o': ['users', '1.0.0', 'list'],
-            'l': 1,
-            'f': [['users', '1.0.0', ['create', 'update']]],
-            'p': {
-              'foo': 'bar'
+        'm': {'s': 'gateway'},
+        'c': {
+          'n': 'rest',
+          'a': {
+            'm': {
+              'v': '1.1.8',
+              'i': 'f678ed17-896f-4431-8e60-d25c61d195ea',
+              'd': '2017-07-21T10:50:17.710670+00:00',
+              't': 2,
+              'p': 'urn:katana:protocol:http',
+              'g': ['127.0.0.1:4444', 'http://10.0.2.15:9999'],
+              'c': '10.0.2.2:64708'
+            },
+            'R': {
+              'b': '{"errors":[{"detail":"User not found","code":"1","status":"404 Not Found"}]}',
+              'v': '1.1',
+              's': '404 Not Found',
+              'h': {
+                'Access-Control-Allow-Origin': '*',
+                'Content-Type': 'application/vnd.api+json'
+              }
+            },
+            'r': {
+              'b': '',
+              'm': 'GET',
+              'v': '1.1',
+              'h': {
+                'HOST': ['localhost:9999'],
+                'ACCEPT': ['*/*'],
+                'USER-AGENT': ['curl/7.51.0']
+              },
+              'u': 'http://localhost:9999/1.0/users/5'
+            },
+            'T': {
+              'e': {
+                'http://10.0.2.15:9999': {
+                  'users': {
+                    '1.0': [{
+                      's': '404 Not Found',
+                      'm': 'User not found',
+                      'c': 404
+                    }]
+                  }
+                }
+              },
+              't': {},
+              'm': {
+                'e': '2017-07-21T10:50:17.762899+00:00',
+                's': '2017-07-21T10:50:17.748366+00:00',
+                'd': '2017-07-21T10:50:17.709640+00:00',
+                'g': ['127.0.0.1:4444', 'http://10.0.2.15:9999'],
+                'v': '1.1.8',
+                'i': 'f678ed17-896f-4431-8e60-d25c61d195ea',
+                'D': 15,
+                'o': ['users', '1.0', 'read'],
+                'l': 1
+              }
             }
           }
         }

--- a/sdk/request.js
+++ b/sdk/request.js
@@ -219,7 +219,8 @@ class Request extends Api {
       new HttpResponse({version, status}),
       new Transport(),
       this._protocol,
-      this._gatewayAddress
+      this._gatewayAddress,
+      null
     );
   }
 

--- a/sdk/request.test.js
+++ b/sdk/request.test.js
@@ -277,9 +277,11 @@ describe('Request', () => {
 
   describe('newResponse()', () => {
     it('should return a Response instance with specified status code and text', () => {
-      const request = new Request(null, null, null, null, null, {}, false, mockHttpRequest);
-      const [statusCode, statusText] = '200 OK'.split(' ');
-      const response = request.newResponse(parseInt(statusCode, 10), statusText);
+      const request = new Request(null, '', '', '', '', {}, false, mockHttpRequest, null, '', '', '', '');
+      const statusCode = 200;
+      const statusText = 'OK';
+
+      const response = request.newResponse(statusCode, statusText);
 
       expect(response).to.be.an.instanceOf(Response);
       expect(response.getHttpResponse().getStatusCode()).equal(statusCode);

--- a/sdk/request.test.js
+++ b/sdk/request.test.js
@@ -277,7 +277,8 @@ describe('Request', () => {
 
   describe('newResponse()', () => {
     it('should return a Response instance with specified status code and text', () => {
-      const request = new Request(null, '', '', '', '', {}, false, mockHttpRequest, null, '', '', '', '');
+      const request = new Request(null, '', '', '', '', {}, false, mockHttpRequest, null,
+        '', '', '', '');
       const statusCode = 200;
       const statusText = 'OK';
 

--- a/sdk/response.js
+++ b/sdk/response.js
@@ -1,6 +1,6 @@
 'use strict';
 
-/* eslint max-params: ["error", 12] */
+/* eslint max-params: ["error", 13] */
 
 const Api = require('./api');
 
@@ -25,17 +25,20 @@ class Response extends Api {
    * @param {Transport} transport
    * @param {string} protocol
    * @param {string} gatewayAddress
+   * @param returnValue
    */
   constructor(
     component, path, name, version, frameworkVersion, variables, debug,
     httpRequest, httpResponse, transport,
-    protocol, gatewayAddress
+    protocol, gatewayAddress,
+    returnValue
   ) {
     super(component, path, name, version, frameworkVersion, variables, debug);
 
     this._httpRequest    = httpRequest;
     this._httpResponse   = httpResponse;
     this._transport      = transport;
+    this._return         = returnValue;
     this._protocol       = protocol;
     this._gatewayAddress = gatewayAddress;
   }
@@ -84,10 +87,14 @@ class Response extends Api {
    * @return {boolean}
    */
   hasReturn() {
-    return this
-      .getServiceSchema(this._name, this._version)
-      .getActionSchema(this._actionName)
-      .hasReturn();
+    return !!this._return;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  getReturn() {
+    return this._return;
   }
 
   /**

--- a/sdk/response.test.js
+++ b/sdk/response.test.js
@@ -18,7 +18,9 @@ describe('Response', () => {
   describe('getTransport()', () => {
     it('should return an instance of the transport', () => {
       const transport = new Transport();
-      const response = new Response(null, null, null, null, null, {}, false, null, null, transport);
+      const response = new Response(
+        null, '', '', '', '', {}, false, null, null, transport, '', '', null
+      );
       expect(response.getTransport()).to.be.an.instanceOf(Transport);
     });
   });

